src/function/execute.rs

```diff
@@ -38,53 +38,87 @@ where
+                ),
+            )
+        });
 
-        // Query was not previously executed, or value is potentially
-        // stale, or value is absent. Let's execute!
-        let database_key_index = active_query.database_key_index;
-        let id = database_key_index.key_index;
-        let value = match Cycle::catch(|| C::execute(db, C::id_to_input(db, id))) {
-            Ok(v) => v,
-            Err(cycle) => {
-                tracing::debug!(
-                    "{database_key_index:?}: caught cycle {cycle:?}, have strategy {:?}",
-                    C::CYCLE_STRATEGY
-                );
-                match C::CYCLE_STRATEGY {
-                    crate::cycle::CycleRecoveryStrategy::Panic => cycle.throw(),
-                    crate::cycle::CycleRecoveryStrategy::Fallback => {
-                        if let Some(c) = active_query.take_cycle() {
-                            assert!(c.is(&cycle));
-                            C::recover_from_cycle(db, &cycle, C::id_to_input(db, id))
-                        } else {
-                            // we are not a participant in this cycle
-                            debug_assert!(!cycle
-                                .participant_keys()
-                                .any(|k| k == database_key_index));
-                            cycle.throw()
+        let mut iteration_count = 0;
+
+        loop {
+            let active_query = zalsa_local.push_query(database_key_index);
+
+            // If we already executed this query once, then use the tracked-struct ids from the
+            // previous execution as the starting point for the new one.
+            if let Some(old_memo) = &opt_old_memo {
+                active_query.seed_tracked_struct_ids(&old_memo.revisions.tracked_struct_ids);
+            }
+
+            // Query was not previously executed, or value is potentially
+            // stale, or value is absent. Let's execute!
+            let mut new_value = C::execute(db, C::id_to_input(db, id));
+            let mut revisions = active_query.pop();
+
+            // If the new value is equal to the old one, then it didn't
+            // really change, even if some of its inputs have. So we can
+            // "backdate" its `changed_at` revision to be the same as the
+            // old value.
+            if let Some(old_memo) = &opt_old_memo {
+                self.backdate_if_appropriate(old_memo, &mut revisions, &new_value);
+                self.diff_outputs(db, database_key_index, old_memo, &mut revisions);
+            }
+
+            // Did the new result we got depend on our own provisional value, in a cycle?
+            if revisions.cycle_heads.contains(&database_key_index) {
+                if let Some(last_provisional) = opt_last_provisional {
+                    if let Some(provisional_value) = &last_provisional.value {
+                        // If the new result is equal to the last provisional result, the cycle has
```

Comment:
I think it is only possible for a memo value to be `None` if its value has been evicted by LRU. I suspect that's practically impossible here, but maybe I need to add some explicit guarantee that LRU eviction of a provisional memo never occurs.

Reply:
Yeah, that's true. I wonder if we can make some use of the knowledge that these memo's can only be reached by tracked functions participating in a cycle. They would have to store the memo value somewhere globally for it to escape and that would require casting the value to `'static`?

Reply:
I suspect this comment was intended for a different thread (the one about creating lots of memos)? Maybe repost it there and delete it here, to make it easier to follow the discussion?

Reply:
It's clearly wrong to consider the cycle converged just because our provisional result got LRU evicted, so I've changed this to an `unwrap` to fail fast in that case, with an explanation comment. Will wait for Niko review before doing more here.